plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/PartitionTable.java

```diff
@@ -207,31 +240,39 @@ private Map<StructLikeWrapper, IcebergStatistics> getStatisticsByPartition(Table
         }
     }
 
-    private RecordCursor buildRecordCursor(Map<StructLikeWrapper, IcebergStatistics> partitionStatistics, List<PartitionField> partitionFields)
+    private RecordCursor buildRecordCursor(Map<StructLikeWrapperWithStructType, IcebergStatistics> partitionStatistics)
     {
-        List<Type> partitionTypes = partitionTypes(partitionFields);
+        List<Type> partitionTypes = partitionTypes();
         List<? extends Class<?>> partitionColumnClass = partitionTypes.stream()
                 .map(type -> type.typeId().javaClass())
                 .collect(toImmutableList());
 
         ImmutableList.Builder<List<Object>> records = ImmutableList.builder();
 
-        for (Map.Entry<StructLikeWrapper, IcebergStatistics> partitionEntry : partitionStatistics.entrySet()) {
-            StructLikeWrapper partitionStruct = partitionEntry.getKey();
+        for (Map.Entry<StructLikeWrapperWithStructType, IcebergStatistics> partitionEntry : partitionStatistics.entrySet()) {
+            StructLikeWrapperWithStructType partitionStruct = partitionEntry.getKey();
             IcebergStatistics icebergStatistics = partitionEntry.getValue();
             List<Object> row = new ArrayList<>();
 
             // add data for partition columns
             partitionColumnType.ifPresent(partitionColumnType -> {
-                BlockBuilder partitionRowBlockBuilder = partitionColumnType.createBlockBuilder(null, 1);
+                BlockBuilder partitionRowBlockBuilder = partitionColumnType.rowType.createBlockBuilder(null, 1);
                 BlockBuilder partitionBlockBuilder = partitionRowBlockBuilder.beginBlockEntry();
                 for (int i = 0; i < partitionColumnTypes.size(); i++) {
-                    io.trino.spi.type.Type trinoType = partitionColumnType.getFields().get(i).getType();
-                    Object value = convertIcebergValueToTrino(partitionTypes.get(i), partitionStruct.get().get(i, partitionColumnClass.get(i)));
+                    io.trino.spi.type.Type trinoType = partitionColumnType.rowType.getFields().get(i).getType();
+                    Object value = null;
+                    for (int j = 0; j < partitionStruct.structType.fields().size(); j++) {
+                        if (partitionStruct.structType.fields().get(j).fieldId() == partitionColumnType.ids.get(i)) {
```

Comment:
Can we use a Map from fieldId -> Type rather than doing the double iteration on the lists?

Reply:
I may be missing something but I failed to create such a map. 
The reason for this inner loop is not to find type(it is found above it) but to assign data  from `partitionStruct` to a correct partition field. The problem is `partitionStruct` comes from reading the files using fileScanTasks and here I am trying to match that with partitioning columns that come from reading the table spec. 
My assumption is that none of this place contains all information this is why I need to match them.
It'd be great if I'm wrong about that, please point it out. 
Tough before I started changing this it was done this way, maybe that was wrong and I should have tried to change it - I doubt this.

Reply:
hmm I may have figure out how to improve this tough